File: app/src/main/java/au/com/codeka/weather/model/Forecast.java
package au.com.codeka.weather.model;

/**
 * POJO that represents a "forecast".
 */
public class Forecast {
  private int offset;
  private String description;
  private String shortDescription;
  private double highTemperature;
  private double lowTemperature;
  private WeatherIcon icon;

  /**
   * Gets the "offset" of this forecast. 0 means it's todays, 1 means tomorrow, 2 the day after etc.
   */
  public int getOffset() {
    return offset;
  }

  public String getDescription() {
    return description;
  }

  public String getShortDescription() {
    return shortDescription;
  }

  public double getHighTemperature() {
    return highTemperature;
  }

  public double getLowTemperature() {
    return lowTemperature;
  }

  public WeatherIcon getIcon() {
    return icon;
  }

  public static class Builder {
    private Forecast forecast;

    public Builder() {
      forecast = new Forecast();
    }

    public Builder setOffset(int offset) {
      forecast.offset = offset;
      return this;
    }

    public Builder setDescription(String description) {
      forecast.description = description;
      return this;
    }

    public Builder setShortDescription(String shortDescription) {
      forecast.shortDescription = shortDescription;
      return this;
    }

    public Builder setHighTemperature(double temp) {
      forecast.highTemperature = temp;
      return this;
    }

    public Builder setLowTemperature(double temp) {
      forecast.lowTemperature = temp;
      return this;
    }

    public Builder setIcon(WeatherIcon icon) {
      forecast.icon = icon;
      return this;
    }

    public Forecast build() {
      return forecast;
    }
  }
}

File: app/src/main/java/au/com/codeka/weather/model/Forecast.kt
package au.com.codeka.weather.model

/**
 * POJO that represents a "forecast".
 */
class Forecast {
  /**
   * Gets the "offset" of this forecast. 0 means it's todays, 1 means tomorrow, 2 the day after etc.
   */
  var offset = 0
    private set
  var description: String? = null
    private set
  var shortDescription: String? = null
    private set
  var highTemperature = 0.0
    private set
  var lowTemperature = 0.0
    private set
  var icon: WeatherIcon? = null
    private set

  class Builder {
    private val forecast: Forecast
    fun setOffset(offset: Int): Builder {
      forecast.offset = offset
      return this
    }

    fun setDescription(description: String?): Builder {
      forecast.description = description
      return this
    }

    fun setShortDescription(shortDescription: String?): Builder {
      forecast.shortDescription = shortDescription
      return this
    }

    fun setHighTemperature(temp: Double): Builder {
      forecast.highTemperature = temp
      return this
    }

    fun setLowTemperature(temp: Double): Builder {
      forecast.lowTemperature = temp
      return this
    }

    fun setIcon(icon: WeatherIcon?): Builder {
      forecast.icon = icon
      return this
    }

    fun build(): Forecast {
      return forecast
    }

    init {
      forecast = Forecast()
    }
  }
}